### PR TITLE
nvme: allow set-features to take input from stdin

### PR DESCRIPTION
**Pull request by jgu-pure**
Fixes: fed9e5ee04cd ("nvme: auto free nvme_dev resource")

### CODE REVIEW

File: nvme.c [hunk 1]
```diff
@@ -5888,6 +5888,8 @@ static int set_feature(int argc, char **argv, struct command *cmd, struct plugin
 							cfg.file, strerror(errno));
 					return -EINVAL;
 				}
+			} else {
+				ffd = STDIN_FILENO;
```
Comment (ikegami-t) [inline nvme.c:5892]: Seems the stdin stream will be closed by the close_file() for cleanup. Is it expected?

Comment (jgu-pure) [reply nvme.c:5892]: Ah, no, sorry. Will fix.

File: nvme.c [hunk 2]
```diff
@@ -5888,9 +5888,10 @@ static int set_feature(int argc, char **argv, struct command *cmd, struct plugin
 							cfg.file, strerror(errno));
 					return -EINVAL;
 				}
+				err = read(ffd, (void *)buf, cfg.data_len);
+			} else {
+				err = read(STDIN_FILENO, (void *)buf, cfg.data_len);
```
Comment (calebsander) [inline nvme.c:5893]: Pre-existing, but doesn't seem like these `void *` casts should be necessary, as that's already the type of `buf`

**Review by calebsander — APPROVED** [review]

File: nvme.c (hunk 1 above)
Comment (igaw) [reply nvme.c]: Why not just dup the `stdin` so we just can use the common close path.

Comment (igaw) [reply nvme.c]: Ignore my comment. I though it's something different. This looks good.

There is still some more cleanup stuff to do with the auto close stuff, e.g. `submit_io` could need some cleanup and I thought we were discussing this here.

Comment (igaw) [reply nvme.c]: Hmm, but I still prefer this here:

```diff
-                       if (strlen(cfg.file)) {
+                       if (strlen(cfg.file))
                                ffd = open(cfg.file, O_RDONLY);
-                               if (ffd <= 0) {
-                                       nvme_show_error("Failed to open file %s: %s",
-                                                       cfg.file, strerror(errno));
-                                       return -EINVAL;
-                               }
-                               err = read(ffd, buf, cfg.data_len);
-                       } else {
-                               err = read(STDIN_FILENO, buf, cfg.data_len);
+                       else
+                               ffd = dup(STDIN_FILENO);
+
+                       if (ffd < 0) {
+                               nvme_show_error("Failed to open file %s: %s",
+                                               cfg.file, strerror(errno));
+                               return -EINVAL;
                        }
+
+                       err = read(ffd, buf, cfg.data_len);
                        if (err < 0) {
                                nvme_show_error("failed to read data buffer from input file: %s",
                                                strerror(errno));
```

Could you update the patch this way? Thanks!

Comment (jgu-pure) [reply nvme.c]: Done!

**Comment by igaw** [comment]
Thanks!

**Pull request merged**